File: chess_app/lib/core/models/engine_game_task.dart
/// „Play it out": a homework item where the student plays a position against
/// the engine until the game ends (docs/PLAN-DOMACI-ZADATAK.md §3, phase 2).
///
/// This file is the app's half of a task both ends read. The server judges
/// again from the moves when the result is recorded
/// (`chess_backend/services/engineGameTask.js`) — nothing here is trusted on
/// its own — but the app has to decide *on its own board* when the game is
/// over, to stop it and say what happened. The two readings are held to one
/// another by a shared fixture, `docs/gates/engine_game_cases.json`.
library;

import 'package:chess/chess.dart' as chess;

import 'package:chess_app/features/exercises/models/exercise_task_words.dart'
    show exercisePieceCount, tablebasePieces;
import 'package:chess_app/services/fen_legality.dart';

import 'drill_outcome.dart';

/// The three goals the first version offers.
enum EngineGameGoal { win, hold, survive }

/// The default and the ceiling for [EngineGameTask.plyCap] — the same numbers
/// the server uses, so a task neither end refuses reads the same limit on
/// both boards.
const int kDefaultEngineGamePlyCap = 200;
const int _maxPlyCap = 600;
const int _maxSurviveMoves = 200;

const List<String> _kEngineLevels = ['lako', 'srednje', 'tesko'];

/// A task as the trainer set it, or nothing: [fromJson] refuses rather than
/// guesses, everywhere it matters. A homework item this app cannot read must
/// not open as a board with an invented goal on it.
class EngineGameTask {
  const EngineGameTask({
    required this.fen,
    required this.side,
    required this.goal,
    this.surviveMoves,
    this.level,
    this.thinkSeconds,
    this.plyCap = kDefaultEngineGamePlyCap,
  });

  /// The position, as the trainer set it.
  final String fen;

  /// The side the student plays; the engine plays the other.
  final chess.Color side;

  final EngineGameGoal goal;

  /// A number of the student's own moves, or null for a game played to its
  /// end. On „hold" and „survive" it is how long to last; on „win" it is
  /// **checkmate in this many moves** — mate on the board by then, or missed.
  final int? surviveMoves;

  /// `'lako'`, `'srednje'` or `'tesko'` — the engine's strength, chosen by the
  /// trainer and travelling on the task. Null leaves the app's own default in
  /// force, the same as the server does for an unset level.
  final String? level;

  /// How long the engine may think, in seconds (1..60), or null for the
  /// app's own default.
  final int? thinkSeconds;

  /// A cap so a game neither side can finish still ends and still counts.
  final int plyCap;

  /// Reads a task from JSON, or refuses it.
  ///
  /// A malformed FEN does not throw here: `chess.Chess.fromFEN` accepts
  /// nonsense silently (measured 17.9.2026 — an "empty board", no exception),
  /// so the check is [isFenLegal] rather than a `try`/`catch` around the
  /// package.
  static EngineGameTask? fromJson(Map<String, dynamic> json) {
    final fenRaw = json['fen'];
    if (fenRaw is! String) return null;
    final fen = fenRaw.trim();
    if (fen.isEmpty || !isFenLegal(fen)) return null;

    final chess.Color side;
    switch (json['side']) {
      case 'w':
        side = chess.Color.WHITE;
        break;
      case 'b':
        side = chess.Color.BLACK;
        break;
      default:
        return null;
    }

    final EngineGameGoal goal;
    switch (json['goal']) {
      case 'win':
        goal = EngineGameGoal.win;
        break;
      case 'hold':
        goal = EngineGameGoal.hold;
        break;
      case 'survive':
        goal = EngineGameGoal.survive;
        break;
      default:
        return null;
    }

    // The number of the student's own moves. `survive` has always needed it;
    // since phase 3a of `docs/PLAN-EXERCISE.md` a `hold` may carry one too —
    // hold the draw for the next N moves. On a `win` the same number means
    // **checkmate in N moves** (the owner's live pass, 19.9.2026): the rules
    // judge it alone, so it has no limit on pieces.
    // `surviveMoves` stays the field's one name on the wire, the same as the
    // server (`chess_backend/services/engineGameTask.js`).
    int? surviveMoves;
    final saidMoves =
        json.containsKey('surviveMoves') && json['surviveMoves'] != null;
    if (goal == EngineGameGoal.survive || saidMoves) {
      final n = _asInt(json['surviveMoves']);
      if (n == null || n < 1 || n > _maxSurviveMoves) return null;
      surviveMoves = n;
    }

    String? level;
    if (json['level'] != null) {
      final s = json['level'].toString();
      if (!_kEngineLevels.contains(s)) return null;
      level = s;
    }

    int? thinkSeconds;
    if (json['thinkSeconds'] != null) {
      final n = _asInt(json['thinkSeconds']);
      if (n == null || n < 1 || n > 60) return null;
      thinkSeconds = n;
    }

    var plyCap = kDefaultEngineGamePlyCap;
    if (json['plyCap'] != null) {
      final n = _asInt(json['plyCap']);
      if (n == null || n < 1 || n > _maxPlyCap) return null;
      plyCap = n;
    }

    return EngineGameTask(
      fen: fen,
      side: side,
      goal: goal,
      surviveMoves: surviveMoves,
      level: level,
      thinkSeconds: thinkSeconds,
      plyCap: plyCap,
    );
  }
}

int? _asInt(dynamic v) {
  if (v is int) return v;
  if (v is String) return int.tryParse(v);
  return null;
}

/// The verdict of an assigned game in progress or just ended.
class EngineGameVerdict {
  const EngineGameVerdict({
    required this.ending,
    required this.outcome,
    required this.goalMet,
    required this.ownMoves,
    required this.needsTablebase,
  });

  /// Null while the game is running.
  final GameEnding? ending;
  final DrillOutcome outcome;
  final bool goalMet;
  final int ownMoves;

  /// True exactly when a „hold" or „survive" game ended at its move target
  /// (`GameEnding.moveTarget`) with seven pieces or fewer on the board. Such
  /// a position is one only a tablebase can judge exactly — the app cannot
  /// ask one, so [goalMet] above is only ever this app's own guess for such a
  /// game, and the caller must wait for the server's word rather than show it
  /// (`docs/PLAN-EXERCISE.md`, phase 3b, decision 1).
  final bool needsTablebase;
}

/// Reads the verdict of an assigned game through [verdictFor] and adds only
/// what a board cannot know: the „survive" target.
///
/// One rule, one home — the five board endings, resignation and the move
/// limit are [verdictFor]'s; this asks `chess.Chess` about none of them a
/// second time. [ownMoves] is not recounted here either: it is the caller's
/// count of the student's own moves played so far (the board's own history
/// starts empty at [EngineGameTask.fen], so plies played and plies allowed
/// mean the same thing to both readers).
EngineGameVerdict engineGameVerdict({
  required EngineGameTask task,
  required chess.Chess game,
  required int ownMoves,
  bool resigned = false,
}) {
  final boardVerdict = verdictFor(
    game,
    task.side,
    plyCap: task.plyCap,
    resigned: resigned,
  );

  var ending = boardVerdict.ending;
  var outcome = boardVerdict.outcome;

  // The board has nothing to say about a move target: it does not know how
  // many of the student's own moves the trainer asked for. Reaching the
  // number is not itself a win or a draw, so the outcome stays undecided even
  // though the game is, from here, over. Since phase 3a of
  // `docs/PLAN-EXERCISE.md` this applies to every goal that carries
  // `surviveMoves`. For a win it is where „checkmate in N moves" is missed:
  // `goalMet` below stays false, and that is final.
  if (ending == null &&
      task.surviveMoves != null &&
      ownMoves >= task.surviveMoves!) {
    ending = GameEnding.moveTarget;
  }

  var goalMet = false;
  if (ending != null) {
    switch (task.goal) {
      case EngineGameGoal.win:
        goalMet = outcome == DrillOutcome.readerWon;
        break;
      case EngineGameGoal.hold:
      case EngineGameGoal.survive:
        // „hold" is anything that is not a loss; „survive" is not losing,
        // either to the end of the game or for the number of the student's
        // own moves the trainer asked for.
        goalMet = outcome != DrillOutcome.readerLost;
        break;
    }
  }

  // With seven pieces or fewer, the position a move target reached is one a
  // tablebase can judge exactly. The app cannot ask one. A win is never
  // asked about: no mate after N moves is a missed goal however won the
  // position still is, and waiting for the server would let it say otherwise.
  final needsTablebase = ending == GameEnding.moveTarget &&
      task.goal != EngineGameGoal.win &&
      exercisePieceCount(game.fen) <= tablebasePieces;

  return EngineGameVerdict(
    ending: ending,
    outcome: outcome,
    goalMet: goalMet,
    ownMoves: ownMoves,
    needsTablebase: needsTablebase,
  );
}

/// The banner over the board: the goal in words and, when the trainer set a
/// number, how many of the student's own moves are left. One home — a goal
/// that carries a number under a sentence that leaves it out is how a student
/// plays on past a limit nobody told them of (the live pass of 19.9.2026).
String engineGameGoalSentence(EngineGameTask task, {required int ownMoves}) {
  final side = task.side == chess.Color.WHITE ? 'White' : 'Black';
  final total = task.surviveMoves;
  if (total == null) {
    return task.goal == EngineGameGoal.win
        ? 'You are $side — win the game'
        : 'You are $side — hold a draw';
  }
  final left = (total - ownMoves).clamp(0, total);
  final moves = '$total ${total == 1 ? 'move' : 'moves'}';
  final ask = task.goal == EngineGameGoal.win
      ? 'checkmate in $moves'
      : 'do not lose for $moves';
  return 'You are $side — $ask · $left left';
}

/// How the game ended, for the dialog that closes it. [endingLabel] names
/// every ending but one: at a move target the words depend on what the number
/// was for — reaching it meets a „hold" and misses a „checkmate in N".
String engineGameEndingWords(EngineGameTask? task, GameEnding ending) {
  if (ending != GameEnding.moveTarget || task == null) {
    return endingLabel(ending);
  }
  final n = task.surviveMoves ?? 0;
  final moves = '$n ${n == 1 ? 'move' : 'moves'}';
  return task.goal == EngineGameGoal.win
      ? 'no checkmate in $moves'
      : 'you were not beaten in $moves';
}

File: chess_backend/services/engineGameTask.js
// engineGameTask.js
//
// „Play it out": the one task in a homework where the student plays a position
// against the engine until the game ends (docs/PLAN-DOMACI-ZADATAK.md §3,
// phase 2). This file holds its shape and its verdict.
//
// **The server judges.** The app plays the game offline and decides on its own
// board when it is over — it has to, to stop the game and say what happened —
// but the recorded result is judged again here from the moves. Nothing else in
// this app lets the client mark its own work (`customPuzzleJudge`, the lesson
// step routes), and a goal a student's device reports as met would be exactly
// that. The two ends are held to one another by a shared fixture,
// `docs/gates/engine_game_cases.json`, which both suites read.

const { Chess } = require('chess.js');

/// `play` has no goal at all (`docs/PLAN-EXERCISE.md`, phase 15): the student
/// plays N moves and the **trainer** says what they were worth. Nothing about
/// it is ever met or missed by the rules, and no tablebase is ever asked.
const GOALS = ['win', 'hold', 'survive', 'play'];

/// The goals that cannot go on without a number: there would be nothing to
/// stop the game by, or nothing to survive.
const NEEDS_MOVES = ['survive', 'play'];

/// Every way an assigned game can be over. The first five are the board's, and
/// mirror `GameEnding` in the app (`lib/core/models/drill_outcome.dart`); the
/// last three cannot be read off a board: the trainer's move limit, the number
/// of moves a „survive" goal asked for, and giving up.
const ENDINGS = [
  'checkmate',
  'stalemate',
  'insufficientMaterial',
  'threefoldRepetition',
  'fiftyMoves',
  'moveLimit',
  'moveTarget',
  'resignation',
];

const DEFAULT_PLY_CAP = 200;
const MAX_PLY_CAP = 600;
const MAX_SURVIVE_MOVES = 200;

/// Reads a task as the trainer set it, or says why it is not one.
///
/// Refused rather than defaulted, everywhere it matters: a position that does
/// not load, a goal nobody defined, or „survive" with no number to survive are
/// all a homework that cannot be answered, and a default would hide which.
/// `level` and `thinkSeconds` are the engine's strength, and they travel **on
/// the task** — the student does not choose the opponent the trainer chose.
function parseEngineGameTask(raw) {
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) {
    return { ok: false, error: 'A game task must be an object.' };
  }
  const { fen, side, goal } = raw;
  if (typeof fen !== 'string' || !fen.trim()) {
    return { ok: false, error: 'A game task needs a position.' };
  }
  let board;
  try {
    board = new Chess(fen);
  } catch (err) {
    return { ok: false, error: `That position is not valid: ${err.message}` };
  }
  if (side !== 'w' && side !== 'b') {
    return { ok: false, error: "The student's side must be 'w' or 'b'." };
  }
  if (!GOALS.includes(goal)) {
    return { ok: false, error: `The goal must be one of ${GOALS.join(', ')}.` };
  }

  // The number of the student's own moves. `survive` has always needed it;
  // since phase 3a of `docs/PLAN-EXERCISE.md` a `hold` may carry one too —
  // hold the draw for the next N moves, judged by the position the game
  // reaches (`goalMetByTablebase`). On a `win` the same number means
  // **checkmate in N moves** (the owner's live pass of 19.9.2026): mate on the
  // board within N, or the goal is missed. The rules judge that alone, so it
  // needs no tablebase and has no limit on pieces.
  //
  // The field keeps its first name on the wire. The plan proposed renaming it
  // `forMoves`; the app in the tree reads and writes `surviveMoves`, and a
  // rename bought a nicer word at the price of a server and an app that
  // disagree until the next phase lands.
  let surviveMoves = null;
  const saidMoves = raw.surviveMoves !== undefined && raw.surviveMoves !== null;
  if (NEEDS_MOVES.includes(goal) || saidMoves) {
    surviveMoves = Number.parseInt(raw.surviveMoves, 10);
    if (!Number.isInteger(surviveMoves) || surviveMoves < 1 || surviveMoves > MAX_SURVIVE_MOVES) {
      return {
        ok: false,
        error: goal === 'survive'
          ? 'Surviving needs a number of moves, from 1.'
          : goal === 'play'
            ? `Playing N moves needs a number from 1 to ${MAX_SURVIVE_MOVES}.`
            : `"For N moves" needs a number from 1 to ${MAX_SURVIVE_MOVES}.`,
      };
    }
  }
  // The engine's strength: the three levels the app already plays at. Left to
  // the app's own default when the trainer did not say.
  const level = raw.level === undefined || raw.level === null ? null : String(raw.level);
  if (level !== null && !['lako', 'srednje', 'tesko'].includes(level)) {
    return { ok: false, error: 'The engine level must be lako, srednje or tesko.' };
  }
  let thinkSeconds = null;
  if (raw.thinkSeconds !== undefined && raw.thinkSeconds !== null) {
    thinkSeconds = Number.parseInt(raw.thinkSeconds, 10);
    if (!Number.isInteger(thinkSeconds) || thinkSeconds < 1 || thinkSeconds > 60) {
      return { ok: false, error: 'The think time must be 1 to 60 seconds.' };
    }
  }

  // A cap so a student cannot be held on one board for ever, and so a game
  // that neither side can finish still ends and still counts.
  let plyCap = DEFAULT_PLY_CAP;
  if (raw.plyCap !== undefined && raw.plyCap !== null) {
    plyCap = Number.parseInt(raw.plyCap, 10);
    if (!Number.isInteger(plyCap) || plyCap < 1 || plyCap > MAX_PLY_CAP) {
      return { ok: false, error: `The move limit must be 1 to ${MAX_PLY_CAP} half-moves.` };
    }
  }

  return {
    ok: true,
    task: {
      fen: board.fen(),
      side,
      goal,
      surviveMoves,
      level,
      thinkSeconds,
      plyCap,
    },
  };
}

/// Replays [moves] from the task's position and says how the game ended and
/// whether the goal was met.
///
/// The order is the same as the app's `verdictFor`: the board first, in a fixed
/// order, then the things a board cannot know — resignation, then the trainer's
/// limit, then the number of moves a „survive" goal asked for. A move the
/// position cannot play is refused, not skipped: `MoveTree.parsePgn` skipping a
/// move in silence is the bug this codebase has paid for most often.
function judgeEngineGame({ task: rawTask, moves, resigned = false }) {
  const parsed = parseEngineGameTask(rawTask);
  if (!parsed.ok) return parsed;
  const task = parsed.task;

  const list = Array.isArray(moves)
    ? moves
    : typeof moves === 'string'
      ? moves.trim().split(/\s+/).filter(Boolean)
      : null;
  if (list === null) return { ok: false, error: 'The moves must be a list.' };
  if (list.length > task.plyCap + 1) {
    return { ok: false, error: 'More moves than the limit allows.' };
  }

  const board = new Chess(task.fen);
  let ownMoves = 0;
  for (const [index, san] of list.entries()) {
    if (typeof san !== 'string') {
      return { ok: false, error: `Move ${index + 1} is not a move.` };
    }
    const mine = board.turn() === task.side;
    try {
      board.move(san);
    } catch (err) {
      return { ok: false, error: `Move ${index + 1} (${san}) cannot be played here.` };
    }
    if (mine) ownMoves++;
  }

  let ending = null;
  if (board.isCheckmate()) ending = 'checkmate';
  else if (board.isStalemate()) ending = 'stalemate';
  else if (board.isInsufficientMaterial()) ending = 'insufficientMaterial';
  else if (board.isThreefoldRepetition()) ending = 'threefoldRepetition';
  else if (board.isDraw()) ending = 'fiftyMoves';
  else if (resigned === true) ending = 'resignation';
  else if (list.length >= task.plyCap) ending = 'moveLimit';
  else if (task.surviveMoves !== null && ownMoves >= task.surviveMoves) ending = 'moveTarget';

  let outcome = 'undecided';
  if (ending === 'checkmate') {
    // After a mate the side to move is the mated one — that is what mate is.
    outcome = board.turn() === task.side ? 'lost' : 'won';
  } else if (ending === 'resignation') {
    outcome = 'lost';
  } else if (['stalemate', 'insufficientMaterial', 'threefoldRepetition', 'fiftyMoves', 'moveLimit'].includes(ending)) {
    outcome = 'drawn';
  }

  // „win" is a win; „hold" is anything that is not a loss, once the game is
  // over; „survive" is not losing, either to the end of the game or for as
  // many of the student's own moves as the trainer asked.
  //
  // „play" has no goal, so it has no answer here: null, however the game
  // ended — mated, mating or stopped at its number — and `needsTrainer` says
  // whose it is. Null and not false: a game nobody has judged must never
  // read as a game that failed.
  const needsTrainer = task.goal === 'play';
  let goalMet = needsTrainer ? null : false;
  if (ending !== null && !needsTrainer) {
    if (task.goal === 'win') goalMet = outcome === 'won';
    else if (task.goal === 'hold') goalMet = outcome !== 'lost';
    else if (task.goal === 'survive') goalMet = outcome !== 'lost';
  }

  // A game that stopped at its move target has no result of its own: nobody
  // won, nobody lost, the trainer's N moves simply ran out. What it reached is
  // the verdict, and with few enough pieces a tablebase knows it exactly. This
  // function stays pure — it says the question needs asking; the caller asks.
  // Until it is answered `goalMet` above is what the rules alone can say: for
  // hold and survive „not lost", which is true.
  //
  // A win is never asked about. „Checkmate in N moves" with no mate on the
  // board after N is missed, however won the position still is — the verdict
  // above is final, and asking would turn a missed mate into a met goal.
  //
  // And „play" is never asked about either: there is no goal for a tablebase
  // to have met.
  const needsTablebase = ending === 'moveTarget'
    && task.goal !== 'win'
    && !needsTrainer
    && pieceCount(board.fen()) <= TABLEBASE_PIECES;

  return {
    ok: true,
    task,
    ending,
    outcome,
    goalMet,
    needsTablebase,
    needsTrainer,
    ownMoves,
    plies: list.length,
    fen: board.fen(),
    moves: list,
  };
}

/// How many pieces stand on the board of [fen], kings included.
function pieceCount(fen) {
  return String(fen).split(' ')[0].replace(/[^a-zA-Z]/g, '').length;
}

/// Seven is as far as any tablebase reaches.
const TABLEBASE_PIECES = 7;

/**
 * Whether the position a game reached meets its goal, given what the
 * tablebase says of it.
 *
 * [category] is the tablebase's word for the side **to move** in [fen]; the
 * student may be either side, so it is turned round when they are not. A
 * cursed win and a blessed loss are draws: the fifty-move rule makes them so,
 * and a game played on would end as one. `wdlOf` throws for 'unknown' and the
 * 'maybe' categories — no outcome is not an outcome, and must not read as a
 * failed homework. Only „hold" and „survive" are ever asked about here.
 */
function goalMetByTablebase({ task, fen, category, wdlOf }) {
  // Loud rather than answered: a win is judged by mate alone
  // (`judgeEngineGame` never says `needsTablebase` for one), so a caller that
  // gets here with a win has skipped that check.
  if (task.goal === 'win') {
    throw new Error('A win is judged by checkmate, never by the tablebase.');
  }
  if (task.goal === 'play') {
    throw new Error('A game with no goal is judged by the trainer, never by the tablebase.');
  }
  const forMover = wdlOf(category);
  const turn = String(fen).split(' ')[1];
  const forStudent = turn === task.side ? forMover : -forMover;
  return forStudent >= -1;
}

module.exports = {
  GOALS,
  TABLEBASE_PIECES,
  pieceCount,
  goalMetByTablebase,
  ENDINGS,
  DEFAULT_PLY_CAP,
  MAX_PLY_CAP,
  parseEngineGameTask,
  judgeEngineGame,
};
